modules/core/src/main/scala/org/tessellation/infrastructure/rewards/Rewards.scala

```diff
@@ -0,0 +1,153 @@
+package org.tessellation.infrastructure.rewards
+
+import cats.data._
+import cats.effect.Async
+import cats.effect.std.Random
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.list._
+import cats.syntax.semigroup._
+
+import scala.collection.immutable.SortedSet
+import scala.concurrent.duration.FiniteDuration
+import scala.math.Integral.Implicits._
+
+import org.tessellation.config.types.RewardsConfig
+import org.tessellation.dag.snapshot.epoch.EpochProgress
+import org.tessellation.domain.rewards._
+import org.tessellation.schema.ID.Id
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.balance.Amount
+import org.tessellation.schema.transaction.{RewardTransaction, TransactionAmount}
+import org.tessellation.security.SecurityProvider
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.cats._
+import eu.timepit.refined.types.numeric._
+import io.estatico.newtype.ops.toCoercibleIdOps
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+object Rewards {
```

Comment:
It is part of the bigger story. There is no way of course to derive `Numeric[A]` because some operations on refined types are not pure. I was thinking about creating a trait which is impure and ops return `Either[ArithmeticException, A]` and then implement that for the numeric types we use so the same as you suggested but in more generic way to keep it DRY. I decided to drop that idea because then we would need to refactor our code to use that ops everwhere instead of for example `Semigroup` in some places for adding 2 refined numbers.

Reply:
Use any kind of abstraction you want, just as long as it's computationally safe and doesn't make a simple arithmetic operation too verbose. Using the abstraction proposed above it took just 6 lines of code to implement startust 10% share.